Ceph: Fix topology logic for drains.

rohantmp opened the pull request:
[test ceph]
<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**
- Fix bug where not cleaning up drain-canaries was leading to a drain deadlock.
- Add an owner reference from the drain-canary to the node.
- As a prerequisite to fixing that, fix some topologyAware issues:
  - In a discussion, I forgot to mention one place where topologyAware changes had to be brought, so parts of the clusterdisruption controller had to be updated to support the newly supported crush types.
  - Label the drain canaries to make them selectable by crush type.



**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [x] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

rohantmp commented:
- [ ] Verified behavior after rebase.

rohantmp commented on pkg/operator/ceph/disruption/nodedrain/reconcile.go:
```diff
@@ -127,19 +127,32 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      k8sutil.TruncateNodeName(fmt.Sprintf("%s-%%s", CanaryAppName), nodeHostnameLabel),
 			Namespace: r.context.OperatorNamespace,
+			OwnerReferences: []metav1.OwnerReference{
```
@leseb I am not using `SetControllerReference` because I want to set the OwnerReference.Controller: false

rohantmp replied:
That works, I'll add that!

leseb commented on cmd/rook/ceph/osd.go:
```diff
@@ -301,37 +301,14 @@ func getLocation(clientset kubernetes.Interface) (string, error) {
-			}
+	topology, invalidLabels := oposd.ExtractRookTopologyFromLabels(nodeLabels)
+	if len(invalidLabels) > 0 {
+		logger.Warningf("ignored invalid node topology labels: %v", invalidLabels)
```
```suggestion
		logger.Warningf("ignored invalid node topology labels: %+v", invalidLabels)
```

rohantmp replied:
The plus adds field names, and an array has no field names to print.

leseb commented on pkg/operator/ceph/cluster/osd/topology.go:
```diff
@@ -30,3 +41,46 @@ var (
 )
+
+// ExtractRookTopologyFromLabels extracts rook topology from labels and returns a map from topology type to value,
+// and a bool indicating if there any invalid labels with a  topology prefix.
```
Where is the bool?

rohantmp replied:
Didn't update the comment when I updated the func signature. Will fix.

leseb commented on pkg/operator/ceph/disruption/clusterdisruption/location.go:
```diff
@@ -160,29 +160,25 @@ func getOSDsForNodes(osdDataList []OsdData, nodeList []*corev1.Node, failureDoma
-			if !ok && secondaryCrushHostname == "" {
-				return nil, fmt.Errorf("could not find the CrushFindResult.Location['%s'] for %s", failureDomainType, osdData.Deployment.ObjectMeta.Name)
+			if !ok {
+				return nil, fmt.Errorf("could not find the CrushFindResult.Location['%s'] for %s", failureDomainType, osdData.Deployment.GetName())
```
```suggestion
				return nil, fmt.Errorf("could not find the CrushFindResult.Location[%q] for %q", failureDomainType, osdData.Deployment.GetName())
```

rohantmp replied:
+1

leseb commented on pkg/operator/ceph/disruption/clusterdisruption/location.go:
```diff
@@ -160,29 +160,25 @@ func getOSDsForNodes(osdDataList []OsdData, nodeList []*corev1.Node, failureDoma
+			nodeFailureDomain, ok := nodeTopologyMap[failureDomainType]
 			if !ok {
-				return nil, fmt.Errorf("could not find the %s label on node %s", failureDomainLabel, node.ObjectMeta.Name)
+				return nil, fmt.Errorf("could not find the %s failure domain on node %s", failureDomainType, node.GetName())
```
```suggestion
				return nil, fmt.Errorf("could not find the %q failure domain on node %q", failureDomainType, node.GetName())
```

leseb commented on pkg/operator/ceph/disruption/clusterdisruption/osd.go:
```diff
@@ -188,6 +191,27 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
+		drainingCanaryList := &appsv1.DeploymentList{}
+		err := r.client.List(context.TODO(), drainingCanaryList, canaryLabels, client.InNamespace(r.context.OperatorNamespace))
+		if err != nil {
+			return fmt.Errorf("could not list canary pods by labels %s: %+v", canaryLabels, err)
```
```suggestion
			return fmt.Errorf("could not list canary pods by labels %q: %+v", canaryLabels, err)
```

leseb commented on pkg/operator/ceph/disruption/clusterdisruption/osd.go:
```diff
@@ -188,6 +191,27 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
+			if time.Since(drainingCanary.GetCreationTimestamp().Time) > time.Minute && drainingCanary.Status.ReadyReplicas < 1 {
+				err := r.client.Delete(context.TODO(), &drainingCanary)
+				if err != nil {
+					logger.Warningf("could not delete canary deployment %s/%s: %+v", drainingCanary.GetName(), drainingCanary.GetNamespace(), err)
```
```suggestion
					logger.Warningf("could not delete canary deployment %q/%q: %+v", drainingCanary.GetName(), drainingCanary.GetNamespace(), err)
```

rohantmp commented on pkg/operator/ceph/disruption/nodedrain/reconcile.go:
```diff
@@ -151,16 +164,24 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
+		for key, value := range topology {
+			objectLabels[key] = value
+		}
+		deploy.ObjectMeta.Labels = objectLabels
```
This is so we can delete canaries by failure domain via labelSelector for cleaning up stale canaries.

Identifying when to delete stale canaries turned out to be a little tricky. Currently, the drain logic deletes the canary for the entire failure domain that is being okayed for draining, and the nodedrain controller brings it back if the node becomes scheduled with OSDs again.

rohantmp commented on pkg/operator/ceph/disruption/nodedrain/reconcile.go:
```diff
@@ -151,16 +164,24 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
+		}
+		topology, _ := osd.ExtractRookTopologyFromLabels(node.GetLabels())
+		for key, value := range topology {
+			objectLabels[key] = value
```
No important reason. I just had a gut feeling that too many selectors could result in a increase in iterations for the Deployment Controller as it identifies which pods match.

Can change it if necessary.

rohantmp commented on pkg/operator/ceph/disruption/nodedrain/reconcile.go:
```diff
@@ -127,19 +127,32 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
+					Kind:               node.Kind,
+					UID:                node.GetUID(),
+					Name:               node.GetName(),
+					Controller:         &controllerBool,
```
+1 will trim

rohantmp replied:
(after testing though. BlockOwnerDeletion has a documented default of false, but Controller does not have a documented default)

rohantmp commented on cmd/rook/ceph/osd.go:
```diff
@@ -301,37 +301,14 @@ func getLocation(clientset kubernetes.Interface) (string, error) {
+		logger.Warningf("ignored invalid node topology labels: %v", invalidLabels)
+	}
+	for topologyType, value := range topology {
+		if topologyType != "host" {
```
Yeah, I was worried about overriding the PVC based hostname with the node hostname. My preference is keeping it explicit that setting the CRUSH host value is not this function's responsibility.

rohantmp commented on pkg/operator/ceph/cluster/osd/topology.go:
```diff
@@ -30,3 +41,46 @@ var (
+			if strings.HasPrefix(labelKey, k8sutil.TopologyLabelPrefix) {
+				s := strings.Split(labelKey, "/")
+				if len(s) != 2 {
+					invalidEncountered = append(invalidEncountered, fmt.Sprintf("%s=%s", labelKey, labelValue))
```
Wanted to have this method as a resuable topology method. We are in fact using it in multiple places, so we'd end up with repeat logs. This way we can reuse it, and discard the info about invalid labels in some places.

rohantmp commented:
> Why use the node as the owner reference? What seems more natural is to add an owner reference to all the CephCluster instances. Then as soon as all the CephCluster CRs are deleted, there is no need to manage the PDBs and the canaries can be removed, right? In the case of a node getting removed, the operator can watch that case and remove the canary, right? The operator will also be monitoring to see if all ceph daemons have been removed from a node and remove the canary from that node as well, right?

For some reason there's no comment button for this one.. 

I think this is a good idea. I'll make the change.

rohantmp commented:
Tested and pushed. Now drain-canaries won't exist for long unless there are OSDs around. No need to set OwnerReferences referrig to multiple CephClusters.

rohantmp commented:
After a conversation with travis based on his comment, removed the node ownerreference from the drain canary and added the operator deployment instead. 
Tested that everything works!

rohantmp commented:
/hold let me just verify the rebase

rohantmp commented:
Done!

rohantmp commented on pkg/operator/ceph/disruption/nodedrain/reconcile.go:
```diff
@@ -151,16 +160,46 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
+		} else {
+
+			operatorDeployment := operatorDeployment
+			controllerBool := true
```
"If true, this reference points to the managing controller."

rohantmp commented on pkg/operator/ceph/disruption/nodedrain/reconcile.go:
```diff
@@ -191,3 +249,35 @@ func newDoNothingContainers(rookImage string) []corev1.Container {
+		return nil, fmt.Errorf("could not get pod %+v: %+v", podKey, err)
+	} else {
+		for _, rsRef := range operatorPod.ObjectMeta.OwnerReferences {
+			if *rsRef.Controller {
```
The pod is  controlled by the replicaset, and the replicaset is controlled by the deployment.

rohantmp commented on pkg/operator/ceph/disruption/clusterdisruption/location.go:
```diff
@@ -169,29 +169,25 @@ func getOSDsForNodes(osdDataList []OsdData, nodeList []*corev1.Node, failureDoma
-				return nil, fmt.Errorf("could not find the %q label on node %q", failureDomainLabel, node.ObjectMeta.Name)
+				return nil, fmt.Errorf("could not find the %q failure domain on node %q", failureDomainType, node.GetName())
 			}
-			if cephClient.IsNormalizedCrushNameEqual(nodeFailureDomain, crushFailureDomain) || cephClient.IsNormalizedCrushNameEqual(secondaryCrushHostname, crushFailureDomain) {
```
Just realized that removing this might break drain-support for non-PVC OSDs where nodeName != labelHostname

rohantmp replied:
@travisn added a tiny commit to tweak this: 
`Ceph: Fix node name checking in the clusterdisruption controller. `

rohantmp replied:
should be good now

rohantmp replied:
(I tested it)

leseb commented:
@travisn your call :-)
